### ISSUE TITLE
How can I reproduce the experiments in the paper?

### ISSUE THREAD

**Issue by chiheonk** [awslabs/adatune#1]
Hi, I'd like to appreciate your great work!

I am trying to reproduce the experiments in Section 6 of your paper. However, it is a bit hard to figure out the exact configuration for each of the experiment (at least for me).

Mostly I was focusing on CIFAR-100 + ResNet-18 case.

For instance, I tried to get the baseline result of `staircase` decay by running the following script.
```shell
python bin/baselines.py --network resnet --dataset cifar_100 \
--optimizer adam --lr 0.0003 --wd 5e-4 \
--lr-scheduler staircase --step-size 60 --lr-decay 0.1
```
As a result, I got 74.xx% accuracy over several runs, which deviates from 76.40% reported in the paper. May I ask for some suggestions for where to look at?

Moreover, I could not find the exact hyperparameter configuration for RTHO and MARTHE. I assumed that the configuration would look like this:
```shell
# RTHO
python bin/rtho.py --network resnet --dataset cifar_100 \ 
--optimizer adam --lr 0.0003 --wd 5e-4 \
--hyper-lr 1e-8 --alpha 0.0 --mu 1.0

# MARTHE
python bin/rtho.py --network resnet --dataset cifar_100 \ 
--optimizer adam --lr 0.0003 --wd 5e-4 \
--hyper-lr 1e-8 --alpha 1e-6 --mu -0.99999
```
Especially, it is a bit hard to understand the role of `mu` when it is negative. The following is from the readme file and the second sentence most confuses me:

> In order to automatically set and adapt mu, set it to any value less than 0.0. You can also pass a value of mu in the range of [0.99, 0.999] if you don't want an adaptive behavior for mu only.

Can you explain further the behavior of the code when `mu` is negative? And, may I ask for an advice (or script) to reproduce experiments as reported in the paper?

Thank you so much for reading!

**Comment by orchidmajumder** [comment]
Hey, thanks for your interest in this project. For `staircase`, you need to use `SGD` with momentum to reproduce our numbers. I see that you are using Adam. With Adam, we found the numbers to be slightly lower than SGD.  May I ask what is the issue you are seeing with RTHO/MARTHE? Is it also related to not matching numbers? @jmikko might help you with parameters for that.

**Comment by chiheonk** [comment]
@orchidmajumder Thank you for the answer! I got better number (77.xx%) with SGD w/ momentum 0.9 and lr 0.1.

From my experience, it was a bit hard to see RTHO converging in a stable way (it was quite sensitive to the randomness generated by CUDA nondeterminism, especially in the beginning) and I could not reproduce the shape of the learning rates with MARTHE and RTHO as in the paper (peaking in early stage -> decaying fast afterward). This is the reason I was asking for the hyperparameter configuration of the experiments, and maybe some know-hows to use them in a stable way.

Thanks!

**Comment by jmikko** [comment]
Hi @chiheonk 
thank you for your interest in this project and sorry for the delayed answer.
You are right, the Readme is not super clear, we will update it soon to make it better and with more details.

As for now, as general rule: any negative value for `mu` enforces the automatic calibration of `mu` during the training. In this case, to ensure the convergence of the method you have to reduce the value of `alpha` (in the experiments we reduced by one order of magnitude). Another useful possibility is to use a fixed value of `mu`, as general rule a value between [0.99, 0.999] (positive) works, but you can also try smaller ones depending on your problem. Concerning RTHO (`mu=1.0`), fixing `alpha=0.0` is correct because you are avoiding the adaptation of the `hyper-lr`, but in this case it is important to reduce the `hyper-lr` itself in order to avoid divergence (the behaviour has to be more conservative).

Finally, I can suggest you also to try the adaptive methods starting with the initial learning rate equals to 0.0. Also in that (disadvantage) scenariothe method is able to retrieve comparable results.
(We will updated the paper soon with further results and more details on the experiments, stay tuned!)

Feel free to contact me also via email for questions or details at donini@amazon.com

Thanks again!

**Comment by chiheonk** [comment]
@jmikko Thank you for your detailed answer! I will definitely try your suggestion.
